libsupport/include/katana/ArrowVisitor.h

```diff
@@ -207,18 +213,30 @@ std::enable_if_t<
         internal::is_visit_arrow_base_t<Arg0>,
         internal::is_visit_arrow_base_t<Args>...>,
     katana::Result<typename std::decay_t<VisitorType>::ReturnType>>
-VisitArrow(VisitorType&& visitor, Arg0 array, Args... args) {
-  return internal::Wrapper<>::template VisitArrowInternal<
+VisitArrow(VisitorType&& visitor, const Arg0& arg0, const Args&... args) {
+  return internal::VisitArrowInternalWrapper<>::template VisitArrowInternal<
       internal::ArrayVisitorBaseType>(
       std::forward<VisitorType>(visitor), std::tuple<>{},
-      std::forward<Arg0>(array), std::forward<Args>(args)...);
+      std::forward<const Arg0>(arg0), std::forward<const Args>(args)...);
```

Comment:
do we want perfect forwarding or value semantics always?

Reply:
The args are always const_reference or pointer, so I'm not sure if it matters?
The visitor needs perfect forwarding

Reply:
If we don't intend to impose a (possible) conversion, I'd rather we stick with perfect forwarding, because that more readily signals to everyone: "just forward these arguments as they were given"

Reply:
The intent of this is to convert a non-const reference to a const reference in the case that someone passes a non-const reference (which is now allowed because of the decay type). Allowing decay types is necessary to enable dereferenced shared_ptrs to be passed as arguments, which is why I included it. 

Here are some options:
 - Enforce constant references through the `std::enable_if` by removing the decay types (shared pointers could go through the other method and I can switch `*arg0` to `*(arg0.get())`) and use perfrect forwarding
 - Use perfect forwarding with decay types and allow non-const references to be passed
 - Keep as is, which will cast non-const refs to const refs

Let me know which you think would be best

Reply:
I'm not familiar with this codepath. I'm guessing the reason that references can't be treated as const references is some template type matching between the initial caller and the final callee?

Given that this code doesn't seem to care too much about how its arguments are managed, would it be possible to just take `const T&` and drop `shared_ptr<T>` arguments (later)?

Anyway, your first two options seem reasonable to me.